### PR TITLE
CDI-38 - Added AttributeRange, rangeMin & rangeMax into the response for the API operation: Retrieve MRCM domain attributes applicable for the given stated parents.

### PR DESCRIPTION
**Pull request by Tyler-Cooksey**
Example Response (which shows attributeRange, rangeMin & rangeMax being returned (if applicable)):

```"attributeRange": [
        {
          "id": "17eec929-cccc-4713-a71d-4d9f422117c5",
          "active": true,
          "referencedComponentId": "1142135004",
          "rangeConstraint": "dec(>#0..)",
          "rangeMin": ">#0",
          "rangeMax": "",
          "attributeRule": "<< 373873005 |Pharmaceutical / biologic product (product)|: [0..*] { [0..1] 1142135004 |Has presentation strength numerator value| > #0 }",
          "ruleStrength": "MANDATORY",
          "contentType": "ALL",
          "dataType": "DECIMAL"
        }...

### CODE REVIEW

File: src/main/java/org/snomed/snowstorm/mrcm/MRCMService.java [hunk 1]
```diff
@@ -112,13 +111,18 @@
...
+
+	private void addExtraConceptMiniFields(final List<AttributeDomain> attributeDomains, final MRCM branchMRCM, final Collection<ConceptMini> attributeConceptMinis) {
+		attributeConceptMinis.forEach(attributeConceptMini -> {
 			attributeConceptMini.addExtraField("attributeDomain",
 					attributeDomains.stream().filter(attributeDomain -> attributeConceptMini.getId().equals(attributeDomain.getReferencedComponentId()))
 							.collect(Collectors.toSet()));
-		}
-
-		return attributeConceptMinis;
+			attributeConceptMini.addExtraField("attributeRange", branchMRCM.getAttributeRanges());
```
Comment (kaicode) [inline src/main/java/org/snomed/snowstorm/mrcm/MRCMService.java:124]: It looks like this adds all attribute ranges to every attributeDomain.
Please only add the attribute ranges applicable to each attributeDomain.
I think the attributeRange.referencedComponentId should match the attributeDomain.referencedComponentId so you could use a filter here.

File: src/main/java/org/snomed/snowstorm/mrcm/model/AttributeRange.java [hunk 2]
```diff
@@ -27,26 +29,18 @@ public AttributeRange(String id, String effectiveTime, boolean active, String re
...
 	public AttributeRange(AttributeRange attributeRange)  {
-		this.id = attributeRange.getId();
-		this.effectiveTime = attributeRange.getEffectiveTime();
-		this.active = attributeRange.isActive();
-		this.referencedComponentId = attributeRange.getReferencedComponentId();
-		this.rangeConstraint = attributeRange.getRangeConstraint();
-		this.attributeRule = attributeRange.getAttributeRule();
-		this.ruleStrength = attributeRange.getRuleStrength();
-		this.contentType = attributeRange.getContentType();
+		this(attributeRange.getId(), attributeRange.getEffectiveTime(), attributeRange.isActive(), attributeRange.getReferencedComponentId(),
```
Comment (kaicode) [inline src/main/java/org/snomed/snowstorm/mrcm/model/AttributeRange.java:32]: Nice reuse of constructors.

File: src/main/java/org/snomed/snowstorm/mrcm/model/AttributeRange.java [hunk 3]
```diff
@@ -27,26 +29,18 @@ public AttributeRange(String id, String effectiveTime, boolean active, String re
...
-		this.effectiveTime = effectiveTime;
-		this.active = active;
-		this.referencedComponentId = referencedComponentId;
-		this.rangeConstraint = rangeConstraint;
-		this.attributeRule = attributeRule;
-		this.ruleStrength = ruleStrength;
-		this.contentType = contentType;
+		this(id, effectiveTime, active, referencedComponentId, rangeConstraint, attributeRule, ruleStrength, contentType);
+		if (dataType != null) {
+			final ConcreteValueRangeConstraint concreteValueRangeConstraint = new ConcreteValueRangeConstraint(rangeConstraint);
```
Comment (kaicode) [inline src/main/java/org/snomed/snowstorm/mrcm/model/AttributeRange.java:40]: Nice reuse, glad you found this 👍

**Review by kaicode — CHANGES_REQUESTED** [review]
A filtering change suggested.

File: src/main/java/org/snomed/snowstorm/mrcm/MRCMService.java (hunk 1 above)
Comment (Tyler-Cooksey) [reply src/main/java/org/snomed/snowstorm/mrcm/MRCMService.java]: Fixed following the comments we talked about in chat:

`we should keep the list of AttributeRanges separate in the response and just filter the list based on the AttributeDomains and content type that has been selected.`

**Pull request closed without merging**